octopoes/octopoes/xtdb/query.py

```diff
@@ -0,0 +1,195 @@
+from dataclasses import dataclass, field
+from typing import List, Optional, Type, Union
+
+from octopoes.models import OOI
+from octopoes.models.path import Direction, Path
+from octopoes.models.types import get_abstract_types, get_relations
+
+
+class InvalidField(ValueError):
+    pass
+
+
+class InvalidPath(ValueError):
+    pass
+
+
+@dataclass
+class Query:
+    """Object representing an XTDB query.
+
+        result_type: The OOI Type being queried: executing the query should yield only this OOI Type.
+
+    Example usage:
+
+    >>> query = Query(Network).where(Network, name="test")
+    >>> query = query.where(Finding, ooi=Network)
+    >>> query.format()
+    '
+    {:query {:find [(pull Network [*])] :where [
+        [ Network :Network/name "test" ]
+        [ Finding :Finding/ooi Network ]
+    ]}}
+    '
+    """
+
+    result_type: Type[OOI]
+
+    _where_clauses: List[str] = field(default_factory=list)
+    _find_clauses: List[str] = field(default_factory=list)
+    _limit: Optional[int] = None
+    _offset: Optional[int] = None
+
+    def where(self, ooi_type: Type[OOI], **kwargs) -> "Query":
+        for field_name, value in kwargs.items():
+            self._where_field_is(ooi_type, field_name, value)
+
+        return self
+
+    def format(self) -> str:
+        return "\n" + self._compile(separator="\n    ") + "\n"
```

Comment:
Why start the string with a newline?

Reply:
Was playing around with some formatting in the tests, this looked pretty easy to read. But the newlines at the start and end perhaps don't add mucht?